cpp/core/memory/MemoryAllocator.cc

```diff
@@ -50,9 +50,13 @@ bool ListenableMemoryAllocator::allocateAligned(uint64_t alignment, int64_t size
 
 bool ListenableMemoryAllocator::reallocate(void* p, int64_t size, int64_t newSize, void** out) {
   int64_t diff = newSize - size;
-  updateUsage(diff);
+  if (diff > 0) {
+    updateUsage(diff);
+  }
   bool succeed = delegated_->reallocate(p, size, newSize, out);
-  if (!succeed) {
+  if (succeed && diff < 0) {
+    updateUsage(diff);
+  } else if (!succeed && diff > 0) {
     updateUsage(-diff);
   }
   return succeed;
```

Comment:
Can we use two different code branches for `diff > 0` and `diff < 0` ? E.g.,

```suggestion
  int64_t diff = newSize - size;
  if (diff >= 0) {
    updateUsage(diff);
    bool succeeded = delegated_->reallocate(p, size, newSize, out);
	if (!succeeded) {
      updateUsage(-diff);
	}
	return succeeded;
  }
  // Code path for diff < 0.
  bool succeeded = delegated_->reallocate(p, size, newSize, out);
  if (succeeded) {
    updateUsage(diff);
  }
  return succeed;
```

Reply:
thanks, changed